beacon-chain/cache/depositsnapshot/deposit_tree.go

```diff
@@ -0,0 +1,102 @@
+package depositsnapshot
+
+import (
+	"crypto/sha256"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/encoding/bytesutil"
+	eth "github.com/prysmaticlabs/prysm/v3/proto/eth/v1"
+)
+
+var (
+	ErrEmptyExecutionBlock = errors.New("empty execution block")
+	ErrInvalidSnapshotRoot = errors.New("snapshot root is invalid")
+	ErrInvalidMixInLength  = errors.New("mixInLength should be greater than 0")
+	ErrInvalidIndex        = errors.New("index should be greater than finalizedDeposits - 1")
+)
+
+// DepositTree is the Merkle tree representation of deposits.
+type DepositTree struct {
+	tree                    MerkleTreeNode
+	mixInLength             uint64 // number of deposits in the tree, reference implementation calls this mix_in_length.
+	finalizedExecutionBlock ExecutionBlock
+}
+
+type ExecutionBlock struct {
+	Hash  [32]byte
+	Depth uint64
+}
+
+// New creates an empty deposit tree.
+func New() *DepositTree {
+	var leaves [][32]byte
+	merkle := create(leaves, DepositContractDepth)
+	return &DepositTree{
+		tree:                    merkle,
+		mixInLength:             0,
+		finalizedExecutionBlock: ExecutionBlock{},
+	}
+}
+
+// getSnapshot returns a deposit tree snapshot.
+func (d *DepositTree) getSnapshot() (DepositTreeSnapshot, error) {
+	if d.finalizedExecutionBlock == (ExecutionBlock{}) {
+		return DepositTreeSnapshot{}, ErrEmptyExecutionBlock
+	}
+	var finalized [][32]byte
+	depositCount, _ := d.tree.GetFinalized(finalized)
+	return fromTreeParts(finalized, depositCount, d.finalizedExecutionBlock), nil
+}
+
+// fromSnapshot returns a deposit tree from a deposit tree snapshot.
+func fromSnapshot(snapshot DepositTreeSnapshot) (DepositTree, error) {
+	if snapshot.depositRoot != snapshot.CalculateRoot() {
+		return DepositTree{}, ErrInvalidSnapshotRoot
+	}
+	tree := fromSnapshotParts(snapshot.finalized, snapshot.depositCount, DepositContractDepth)
+	return DepositTree{
+		tree:                    tree,
+		mixInLength:             snapshot.depositCount,
```

Comment:
Do some validation on deposit count before initializing

Reply:
check how deep the tree is (should have a depth), then number of deposits allowed is 2^depth